### ISSUE TITLE
[C++][Python] pyarrow.ChunkedArray.combine_chunks is slow

### ISSUE THREAD

**Issue by spenczar** [apache/arrow#37318]
### Describe the enhancement requested

Chunked arrays are common in Pyarrow, since Tables are build on chunked arrays. Working with the underlying data requires combining chunks.

Combining chunks is surprisingly slow. In my application, I need to work with actual arrays, but my code spends a significant amount of its time calling combine_chunks even when there is just one chunk.

I have some informal benchmarking I did in an IPython session, which shows an improvement from ~150 microseconds to ~300 nanoseconds for an array of 1 million floats:

```python
In [1]: import pyarrow as pa

In [2]: import numpy as np

In [3]: data = pa.array(np.random.random(1_000_000))

In [4]: chunked_array = pa.chunked_array([data])

In [5]: %timeit chunked_array.combine_chunks()
154 µs ± 1.66 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [6]: def fast_combine_chunks(chunked_array):
   ...:     if chunked_array.num_chunks == 1:
   ...:         return chunked_array.chunk(0)
   ...:     return chunked_array.combine_chunks()
   ...:

In [7]: %timeit fast_combine_chunks(chunked_array)
290 ns ± 2.24 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

### Component(s)

Python

**Comment by lidavidm** [comment]
Hmm, interesting. Ideally on the C++ side we would change concat_arrays to not allocate a new array if there's only one chunk. But that would actually be a breaking change, since that behavior has been used for years as the only way to copy an array (sometimes you explicitly want a copy for memory usage reasons). So we're in a bit of a bind on the C++ side.

**Comment by lidavidm** [comment]
For reference: #26623, #30503

**Comment by js8544** [comment]
> Hmm, interesting. Ideally on the C++ side we would change concat_arrays to not allocate a new array if there's only one chunk. But that would actually be a breaking change, since that behavior has been used for years as the only way to copy an array (sometimes you explicitly want a copy for memory usage reasons). So we're in a bit of a bind on the C++ side.

I agree that we shouldn't change `Concatenate` because that would be breaking. But I think this particular optimization can happen within `combine_chunks` without changing the behavior of `Concatenate`. C++'s `table` has [the same optimization](https://github.com/apache/arrow/blob/main/cpp/src/arrow/table.cc#L546) as the one OP suggested.

**Comment by jorisvandenbossche** [comment]
I agree it sounds OK to change this just for the `ChunkedArray.combine_chunks` method (not for `concat_arrays` used under the hood there), especially if this is making it consistent with the `Table.combine_chunks` behaviour.

Sidenote: the end goal here is to get an Array object (so you still need to do `.chunk(0)` after combining). I am wondering if we want to add a method just for converting ChunkedArray into Array.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
